---
title: "Moving Pictures Magnitude Reappearance"
author: "Daniel J. Park"
date: "7/16/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(ggplot2)
library(patchwork)
library(dplyr)
library(zoib)
require(foreign)
require(Hmisc)
library(microbiome)
source("../mp_analysis_functions.R")
```

```{r}
# f4.metadata : metadata sorted by time
# otu.counts.f4 : OTU counts
# otu.relabs.f4 : OTU relative abundances
# otu.clrabs.f4 : OTU clr-transformed abundances
# median.otu.relabs.f4 : OTU's median relative abundance across all samples
# median.otu.clrabs.f4 : OTU's median clr-transformed abundance across all samples
# coverage.f4 : the total number of reads by sample
load("../mp_F4_data.Rdata")
load("../mp_M3_data.Rdata")
load("../dethlefsen_relman.Rdata")
load("../colonization_analysis/bigReappTab.Rdata")
load("../otu_tables_MP.Rdata")
```

```{r}
# set <- sample(x = 1:500, size = 1)
set <- 63
sim.dataset <- as.matrix(read.table(file = paste("../SimSets_MP_n2_t120", "/set", sprintf("%04d", set), ".txt", sep = "")))

# sim.otutab.f4 <- as.data.frame(sim.dataset) %>%
#   dplyr::select(contains("F4"))
sim.otutab.m3 <- as.data.frame(sim.dataset) %>%
  dplyr::select(contains("M3"))

sim.reappeared.relabs.tab <- format_reappearance_data(otutab = sim.otutab.m3, 
                                                            otu.rarity = avg.asin.abnd(sim.otutab.m3), 
                                                            subj.id = "M3") %>%
  filter(reappeared == 1)

summary(sim.reappeared.relabs.tab)

ggplot(sim.reappeared.relabs.tab, aes(x = otu.rarity, y = reapp.abnd)) +
  geom_point(aes(color = subj.id),
             size = .75) + 
  labs(title = "OTU Reappeared Relative Abundances",
       subtitle = paste("Simulated Moving Pictures dataset set", sprintf("%04d", set), sep = ""),
       x = "average within-subject arcsin Hellinger abundance",
       y = "relative abundance") +
  ylim(0,0.2) +
  xlim(0, 0.25)

```

# Predict Reappearance Relative Abundances

```{r}
# dr.mp.reapp.tab$subjID <- as.factor(dr.mp.reapp.tab$subjID)
# reappeared.relabs.tab <- dr.mp.reapp.tab %>%
#   filter(y > 0)
reappeared.relabs.tab <- format_reappearance_data(otutab = rel.otutab.m3, 
                                                        otu.rarity = avg.asin.abnd(rel.otutab.m3), 
                                                        subj.id = "M3") %>%
  filter(reappeared == 1)

summary(reappeared.relabs.tab)

ggplot(reappeared.relabs.tab, aes(x = otu.rarity, y = reapp.abnd)) +
  geom_point(aes(color = subj.id),
             size = .75) + 
  labs(title = "OTU Reappeared Relative Abundances",
       subtitle = "Observed Moving Pictures dataset",
       x = "average within-subject arcsin Hellinger abundance",
       y = "relative abundance") +
  ylim(0,0.2) +
  xlim(0, 0.25)

```

```{r}
library(glmmTMB)

reapp.glmmTMB <- glmmTMB(y ~ otu.rarity + (otu.rarity|subjID), 
                         data = reappeared.relabs.tab, 
                         family = beta_family)

sim.reapp <- simulate(reapp.glmmTMB)
reappeared.relabs.tab$sim <- sim.reapp[ ,1]
head(reappeared.relabs.tab)

ggplot(reappeared.relabs.tab, aes(x = otu.rarity, y = sim)) +
  geom_point(aes(color = subjID),
             size = .75) + 
  labs(title = "Simulated reappeared relative abundances",
       x = "OTU mean arcsin Hellinger-abundance (within subject)",
       y = "") +
  ylim(0, 0.15)
```

```{r}
summary(reappeared.relabs.tab)

# this is the proportion of true reappearances larger than
# the greatest simulated reappearance
reappeared.relabs.tab %>%
  mutate(large.reapp = if_else(y > max(sim), 1, 0)) %>%
  summarise(mean(large.reapp))
  
# reappeared.relabs.tab %>%
#   filter(y > 0.005) %>%
#   distinct(otu.id, otu.rarity, subjID) %>%
#   summarise(range(otu.rarity))

# model summary
summary(reapp.glmmTMB)
# random effects
ranef(reapp.glmmTMB)
```

```{r}
# predicting with a 'new' dataset

data.list <- list(list(otu.relabs.f4, "F4"),
                  list(otu.relabs.m3, "M3"),
                  list(dr.D.rel.otutab, "D"),
                  list(dr.E.rel.otutab, "E"),
                  list(dr.F.rel.otutab, "F"))
new.reapp.df <- data.frame(otu.rarity = numeric(),
                           subjID = character())

for (subj in data.list) {
  subj.otutab <- subj[[1]]
  subj.id <- subj[[2]]
  
  subj.otu.rarity <- otu.avg.asin(subj.otutab)
  subj.df <- data.frame(otu.rarity = subj.otu.rarity,
                        subjID = rep(subj.id, length(subj.otu.rarity)))
  new.reapp.df <- rbind(new.reapp.df, subj.df)
}

pred.reapp.abnd.1 <- predict(object = reapp.glmmTMB, newdata = new.reapp.df, allow.new.levels = FALSE, type = "response")
pred.reapp.abnd.2 <- predict(object = reapp.glmmTMB, newdata = new.reapp.df, allow.new.levels = TRUE)

head(cbind(pred.reapp.abnd.1, pred.reapp.abnd.2))

new.reapp.df$pred.1 <- pred.reapp.abnd.1

ggplot(new.reapp.df, aes(x=otu.rarity, y=pred.1, color=subjID)) +
  geom_point() +
  xlim(c(0,0.25)) +
  ylim(c(0,0.15))

summary(pred.reapp.abnd.1)
```


```{r}
reapp.zoib <- zoib(y ~ otu.rarity | otu.rarity | subjID,
                   data = reappeared.relabs.tab,
                   joint = FALSE,
                   zero.inflation = FALSE,
                   one.inflation = FALSE,
                   random = 12,
                   EUID = 1:nrow(reappeared.relabs.tab))
```


```{r}
reapp.relabs.f4 <- format_reappearance_data(otutab = otu.relabs.f4,
                                            otuRarity = median.otu.clrabs.f4,
                                            subjID = "F4")

head(reapp.relabs.f4); tail(reapp.relabs.f4); summary(reapp.relabs.f4); dim(reapp.relabs.f4)
```

```{r}
reapp.relabs.m3 <- format_reappearance_data(otutab = otu.relabs.m3,
                                            otuRarity = median.otu.clrabs.m3,
                                            subjID = "M3")

head(reapp.relabs.m3); tail(reapp.relabs.m3); summary(reapp.relabs.m3); dim(reapp.relabs.m3)
```

The large raw counts are coming from samples with large read counts, which is why we're seeing a similar distribution of reappearing relative abundances. I think the best model to use is a zero-inflated negative binomial regression $y \sim x_1 + x_2$ where $x_1$ is the within-subject OTU median abundance and $x_2$ is log(sample read depth). Or a non-zero-inflated negative binomial regression regressing the count on within-subject OTU median abundance, log(sample read depth), and time since last sample. 

```{r}
ggplot(reapp.relabs.f4) +
  geom_point(aes(x = otu.rarity, y = y)) +
  labs(title = "Subject F4", 
       x = "OTU median clr abundance",
       y = "Reappeared Relative Abundance (zeros included)")

# ggplot(reapp.relabs.f4) +
#   geom_point(aes(x = sample.read.depth, y = y)) +
#   labs(title = "Subject F4", y = "Reappeared Relative Abundances")

d <- subset(reapp.relabs.f4, y < 0.04)

observed.reapp.f4 <- ggplot(d) +
  geom_point(aes(x = median.abundance, y = y)) +
  labs(title = "F4 Observed Reappearances", y = "Relative Abundance")

# ggplot(d) +
#   geom_point(aes(x = sample.read.depth, y = y)) +
#   labs(title = "Subject F4 without outliers", y = "Reappeared Relative Abundances")

```

```{r}
observed.reapp.m3 <- ggplot(reapp.relabs.m3) +
  geom_point(aes(x = otu.rarity, y = y)) +
  labs(title = "M3 Observed Reappearances", 
       x = "OTU median clr abundance",
       y = "Reappeared Relative Abundance (zeros included)")

observed.reapp.m3
```


```{r}
load("m3_reapp_rel_ab_ZIBR.Rda")
load("f4_reapp_rel_ab_ZIBR.Rda")
library(cowplot)
```

```{r}
zibr.coefs.f4 <- summary(f4.reapp.relabs.ZIBR$coeff)$statistics[ ,1]
zibr.coefs.m3 <- summary(m3.reapp.relabs.ZIBR$coeff)$statistics[ ,1]

# SIMULATE THE ZOIB MODEL
d$sim.y <- sim.zoib(f4.reapp.relabs.ZIBR, 
                    beta.mean.coefs = zibr.coefs.f4[1:3], 
                    infl.coefs = logit(0.54), #zibr.coefs.f4[4], 
                    beta.disp.coefs = zibr.coefs.f4[5:6])

reapp.relabs.m3$sim.y <- sim.zoib(m3.reapp.relabs.ZIBR,
                                  beta.mean.coefs = zibr.coefs.m3[1:3],
                                  infl.coefs = logit(0.54), #zibr.coefs.m3[4],
                                  beta.disp.coefs = zibr.coefs.m3[5:6])
```

```{r}
simulated.reapp.f4 <- ggplot(d) +
  geom_point(aes(x = median.abundance, y = sim.y)) +
  ylim(0, max(d$y)) +
  labs(title = "F4 Simulated", y = "Relative Abundance")

reapp.grid.f4 <- plot_grid(observed.reapp.f4, simulated.reapp.f4, ncol = 2)

reapp.grid.f4

save_plot("reapp_rel_abs_f4.png", reapp.grid.f4, ncol = 2)

sum(d$y > 0.005)
mean(d$y > 0.005)
```

```{r}
simulated.reapp.m3 <- ggplot(reapp.relabs.m3) +
  geom_point(aes(x = median.abundance, y = sim.y)) + 
  ylim(0, max(reapp.relabs.m3$y)) +
  labs(title = "M3 Simulated", y = "Relative Abundance")

reapp.grid.m3 <- plot_grid(observed.reapp.m3, simulated.reapp.m3, ncol = 2)

reapp.grid.m3

save_plot("reapp_rel_abs_m3.png", reapp.grid.m3, ncol = 2)

sum(reapp.relabs.m3$y > 0.05)
mean(reapp.relabs.m3$y > 0.05)
```

```{r}
mean(d$y==0)
mean(d$sim.y==0) # much higher simulated 0 inflation than observed

mean(reapp.relabs.m3$y==0)
mean(reapp.relabs.m3$sim.y==0) # same for subject M3
```

## Dethlefsen and Relman data

```{r}
produce.reapp.tab <- function(otu.tab, med.clr, subj.ID) {
  reapp.prob <- reappearance_probabilities(otu.tab)
  reapp.tab <- data.frame(reapp.prob = reapp.prob,
                          median.clr = med.clr,
                          subj.ID = rep(subj.ID, length(med.clr)))
  return(reapp.tab)
}
```

```{r}
big.reapp.tab <- rbind(produce.reapp.tab(dr.D.otutab, dr.D.med.clr, "D"),
                       produce.reapp.tab(dr.E.otutab, dr.E.med.clr, "E"),
                       produce.reapp.tab(dr.F.otutab, dr.F.med.clr, "F"),
                       produce.reapp.tab(otu.counts.f4, median.otu.clrabs.f4, "F4"),
                       produce.reapp.tab(otu.counts.m3, median.otu.clrabs.m3, "M3"))

big.reapp.plot <- ggplot(big.reapp.tab, aes(x = median.clr, y = reapp.prob, color = factor(subj.ID))) +
  geom_point(size = .8) +
  labs(y = "Probability of reappearance from absence",
       x = "OTU median clr abundance",
       title = "Probability of OTU reappearance",
       subtitle = "Dethlefsen and Relman (2011) and Caporaso et al.",
       color = "Subject ID") 

big.reapp.plot
```

```{r}
big.reapp.plot + 
  geom_smooth(method = "glm", se = FALSE, method.args = list(family=binomial))
```

## Asymptotic regression

```{r}
mp.F4.reapp.tab <- produce.reapp.tab(otu.counts.f4, median.otu.clrabs.f4, "F4")
```


# Phylogenetic patterns to reappearances

```{r}
library(MicrobeDS)
library(cluster)
library(phyloseq)
```

```{r}
data('MovingPictures')
mp.tree <- MovingPictures@phy_tree
# mp.tree$Nnode

unique.otu.labels <- unique(c(rownames(otu.counts.f4), rownames(otu.counts.m3)))

pruned.mp.tree <- prune_taxa(unique.otu.labels, mp.tree)

mp.tree.dist <- cophenetic(pruned.mp.tree)

# SELECTING THE OTUS WITH HIGHER REL. ABS. REAPPEARANCES
high.reapp.otus.f4 <- reapp.relabs.f4 %>%
  filter(y > 0.005) %>%
  dplyr::select(otu_ids) %>%
  distinct()

high.reapp.otus.m3 <- reapp.relabs.m3 %>%
  filter(y > 0.04) %>%
  dplyr::select(otu_ids) %>%
  distinct()

for (nClus in c(5,10,20)) {
  pam.obj <- pam(mp.tree.dist, nClus)
  cluster.labels <- pam.obj$clustering
  
  df.f4 <- data.frame(cluster.labels[which(names(cluster.labels) %in% high.reapp.otus.f4$otu_ids)])
  colnames(df.f4) <- c("cluster")
  barchart.f4 <- ggplot(df.f4, aes(x = cluster)) +
    geom_bar(stat = "count") +
    labs(title = "F4", subtitle = "High-relative abundance reappearance OTUs")
  
  df.m3 <- data.frame(cluster.labels[which(names(cluster.labels) %in% high.reapp.otus.m3$otu_ids)])
  colnames(df.m3) <- c("cluster")
  barchart.m3 <- ggplot(df.m3, aes(x = cluster)) +
    geom_bar() +
    labs(title = "M3", subtitle = "High-relative abundance reappearance OTUs")

  show(barchart.f4 + barchart.m3)
}


```


# Raw Abundances of reappearance

Set up a regression problem for predicting the magnitude of a reappearing OTU count (the positive count for OTU $k$ at time $t$ given a zero count for OTU $k$ at time $t-1$).

```{r}
mag.reapp.f4 <- format_reappearance_data(otutab = otu.counts.f4, 
                                         otu.median.abs = median.otu.clrabs.f4,
                                         read.depths = coverage.f4,
                                         otu_id = names(median.otu.clrabs.f4))

head(mag.reapp.f4); tail(mag.reapp.f4); summary(mag.reapp.f4); dim(mag.reapp.f4)

```


For a pretty rare OTU, is there a difference in the distributions of sample depth (total read count) for reappearance and non-reappearance samples?

```{r}
rare.otus <- c("174950", "183207", "495451")
median.otu.relabs.f4[which(names(median.otu.relabs.f4) == "174950" )]
median.otu.relabs.f4[which(names(median.otu.relabs.f4) == "183207" )]
median.otu.relabs.f4[which(names(median.otu.relabs.f4) == "495451" )]

rare.reapps <- subset(mag.reapp.f4, otu_ids %in% rare.otus)
dim(rare.reapps)

rare.reapps <- rare.reapps %>% 
  mutate(isReappearance = case_when(y == 0 ~ "non-reappearance", y > 0 ~ "reappearance"))

rare.reapps

ggplot(rare.reapps) +
  geom_boxplot(aes(x = isReappearance, y = sample.read.depth, fill = as.factor(otu_ids)))
 
```


The frequency table looks like the distribution of reappearance counts is highly zero-inflated.

```{r}
ggplot(mag.reapp.f4, aes(y)) +
  geom_histogram(binwidth = 1)

ggplot(mag.reapp.f4, aes(x=median.abundance, y)) +
  geom_point()

ggplot(mag.reapp.f4, aes(x=sample.read.depth, y)) +
  geom_point()

ggplot(mag.reapp.f4, aes(x=time.diff, y)) +
  geom_point()

as.data.frame(table(mag.reapp.f4$y))
```


0) (Zero-inflated?) Negative Binomial Regression

```{r}
require(pscl)
require(MASS)
require(boot)
```

A first pass zero-inflated negative binomial model predicting the reappearance count.

```{r}
m1 <- zeroinfl(y ~ median.abundance | median.abundance,
               data = mag.reapp.f4, dist = "negbin")

summary(m1)
m1$coefficients
m1$theta
```

Compared to a model with just an intercept, m1 is statistically significant.
```{r}
m0 <- update(m1, . ~ 1)

pchisq(2 * (logLik(m1) - logLik(m0)), df = 2, lower.tail=FALSE)
```

Model 2 is not zero-inflated.

```{r}
m2 <- glm.nb(y ~ median.abundance, data = mag.reapp.f4)

summary(m2)
m2$coefficients
m2$theta
```


Compare the zero-inflated vs. not-inflated negative binomial regression models. Is the Vuong test an appropriate method?
```{r}
vuong(m1, m2)
```

Add time between consecutive samples into the mix, but only an intercept for zero-inflation binomial with logit link.
```{r}
m3 <- zeroinfl(y ~ median.abundance + time.diff | 1,
               data = mag.reapp.f4, dist = "negbin")

summary(m3)
```

The negative binomial part depends on abundance ranking and log sample depth.

```{r}
m4 <- zeroinfl(y ~ median.abundance + log(sample.read.depth) | 1,
               data = mag.reapp.f4, dist = "negbin")

summary(m4)
```


No zero-inflation, but include all the covariates in the negative binomial regression.

```{r}
m5 <- glm.nb(y ~ median.abundance + log(sample.read.depth) + time.diff, data = mag.reapp.f4)

summary(m5)
```

Compare predictions from zero-inflated vs. not-inflated negative binomial regression models.

```{r}
new.data <- mag.reapp.f4 %>% dplyr::select( median.abundance, sample.read.depth, time.diff)

new.data$pred.m1 <- predict_reappearance(X_count = list(new.data$median.abundance), 
                                         X_zero = list(new.data$median.abundance), 
                                         negbin_model = m1, zero_inflated = TRUE, seed=0)

new.data$pred.m2 <- predict_reappearance(X_count = list(new.data$median.abundance), 
                                         negbin_model = m2, zero_inflated = FALSE, seed=0)

new.data$pred.m3 <- predict_reappearance(X_count = list(new.data$median.abundance, new.data$time.diff), 
                                         negbin_model = m3, zero_inflated = TRUE, seed=0)

new.data$pred.m4 <- predict_reappearance(X_count = list(new.data$median.abundance, log(new.data$sample.read.depth)),
                                         negbin_model = m4, zero_inflated = TRUE, seed=0)

new.data$pred.m5 <- predict_reappearance(X_count = list(new.data$median.abundance, log(new.data$sample.read.depth), new.data$time.diff),
                                         negbin_model = m5, zero_inflated = FALSE, seed=0)
```

```{r}
ggplot(mag.reapp.f4)+
  geom_point(aes(x = median.abundance, y = y), color="purple") +
  labs(x = "OTU median clr-transformed abundance", y = "Observed Reappearance Counts")

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.m1), color="sienna") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (model 1)") +
  ylim(0, max(mag.reapp.f4$y))

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.m2), color="gold") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (model 2)") +
  ylim(0, max(mag.reapp.f4$y))

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.m3), color="seagreen") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (model 3)") +
  ylim(0, max(mag.reapp.f4$y))

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.m4), color="turquoise") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (model 4)") +
  ylim(0, max(mag.reapp.f4$y))


ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.m5), color="tomato") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (model 5)") +
  ylim(0, max(mag.reapp.f4$y))


as.data.frame(table(mag.reapp.f4$y))
as.data.frame(table(new.data$pred.m1))
as.data.frame(table(new.data$pred.m2))
as.data.frame(table(new.data$pred.m3))
as.data.frame(table(new.data$pred.m4))
as.data.frame(table(new.data$pred.m5))

```


# Raw Abundance of Reappearance Counts without high outliers

```{r}
# don't include any observed reappearance counts above 1000
nonoutlier.reapp.counts <- subset(mag.reapp.f4, y < 1000)

head(nonoutlier.reapp.counts)
```

```{r}
new.m1 <- update(m1, data = nonoutlier.reapp.counts)
new.m2 <- update(m2, data = nonoutlier.reapp.counts)
new.m3 <- update(m3, data = nonoutlier.reapp.counts)
new.m4 <- update(m4, data = nonoutlier.reapp.counts)
new.m5 <- update(m5, data = nonoutlier.reapp.counts)

new.data$pred.new.m1 <- predict_reappearance(X_count = list(new.data$median.abundance), 
                                         X_zero = list(new.data$median.abundance), 
                                         negbin_model = new.m1, zero_inflated = TRUE, seed=0)

new.data$pred.new.m2 <- predict_reappearance(X_count = list(new.data$median.abundance), 
                                         negbin_model = new.m2, zero_inflated = FALSE, seed=0)

new.data$pred.new.m3 <- predict_reappearance(X_count = list(new.data$median.abundance, new.data$time.diff), 
                                         negbin_model = new.m3, zero_inflated = TRUE, seed=0)

new.data$pred.new.m4 <- predict_reappearance(X_count = list(new.data$median.abundance, log(new.data$sample.read.depth)),
                                         negbin_model = new.m4, zero_inflated = TRUE, seed=0)

new.data$pred.new.m5 <- predict_reappearance(X_count = list(new.data$median.abundance, log(new.data$sample.read.depth), new.data$time.diff),
                                         negbin_model = new.m5, zero_inflated = FALSE, seed=0)

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.new.m1), color="sienna") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (new model 1)") +
  ylim(0, max(mag.reapp.f4$y))

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.new.m2), color="gold") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (new model 2)") +
  ylim(0, max(mag.reapp.f4$y))

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.new.m3), color="seagreen") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (new model 3)") +
  ylim(0, max(mag.reapp.f4$y))

ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.new.m4), color="turquoise") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (new model 4)") +
  ylim(0, max(mag.reapp.f4$y))


ggplot(new.data) +
  geom_point(aes(x = median.abundance, y = pred.new.m5), color="tomato") +
  labs(x = "OTU median clr-transformed abundance", y = "Simulated Reappearance Counts (new model 5)") +
  ylim(0, max(mag.reapp.f4$y))

as.data.frame(table(new.data$pred.new.m1))
as.data.frame(table(new.data$pred.new.m2))
as.data.frame(table(new.data$pred.new.m3))
as.data.frame(table(new.data$pred.new.m4))
as.data.frame(table(new.data$pred.new.m5))

```


```{r}
ggplot(new.data) +
  geom_histogram(aes(pred.m5), binwidth = 1)

ggplot(new.data) +
  geom_histogram(aes(pred.new.m5), binwidth = 1)

```


# Ordinal logistic regression

```{r}
positive.reappearances.f4 <- subset(mag.reapp.f4, y > 0)

positive.y <- positive.reappearances.f4$y
# singeton
# doubleton
# 3 or 4
# 5 or greater
positive.reappearances.f4$y.factor <- factor(case_when(positive.y == 1 ~ 1,
                                                       positive.y == 2 ~ 2,
                                                       positive.y == 3 | positive.y == 4 ~ 3,
                                                       positive.y >= 5 & positive.y <= 10 ~ 5,
                                                       positive.y > 10 ~ 11
                                                       ), 
                                             ordered = TRUE,
                                             levels = c(1, 2, 3, 5, 11))
as.data.frame(table(positive.reappearances.f4$y.factor))

```


The magnitude of reappearance looks associated the OTU's median clr-transformed abundance and the sample's log total read count.
```{r}
ggplot(positive.reappearances.f4, aes(x = median.abundance, y = y.factor , fill = y.factor)) +
  geom_boxplot() +
  labs(x = "OTU median clr-transformed abundance", y = "Reappearance Count Class")

ggplot(positive.reappearances.f4, aes(x = log(sample.read.depth), y = y.factor , fill = y.factor)) +
  geom_boxplot() +
  labs(x = "log Sample Read Depth", y = "Reappearance Count Class")

ggplot(positive.reappearances.f4, aes(x = time.diff, y = y.factor , fill = y.factor)) +
  geom_boxplot() + 
  labs(x = "Difference in Time between Reappearance and Previous Sample", y = "Reappearance Count Class")
```



```{r}
m.olr <- polr(y.factor ~ median.abundance + log(sample.read.depth), 
              data = positive.reappearances.f4, Hess = TRUE)

summary(m.olr)
```


$logit(\hat{P}(Y \leq 1)) = 11.9915 - 0.7849*med.abund - 1.2671*log(read.depth)$
$logit(\hat{P}(Y \leq 2)) = 12.9508 - 0.7849*med.abund - 1.2671*log(read.depth)$
$logit(\hat{P}(Y \leq 3)) = 14.0641 - 0.7849*med.abund - 1.2671*log(read.depth)$

Approximate the t-distribution with a standard normal (good for large sample sizes) and calculate p-values for the coefficients. All the coefficients have significant p-values.
```{r}
olr.ctable <- coef(summary(m.olr))

olr.pvals <- pnorm(abs(olr.ctable[ , "t value"]), lower.tail = FALSE) * 2
(olr.ctable <- cbind(olr.ctable, "p value" = olr.pvals))
```


Approximate CIs with profiling the likelihood function and by assuming a normal distribution. Neither method gives a CI for a coefficient that contains 0.
```{r}
(olr.profiled.ci <- confint(m.olr)) # profiled CIs
confint.default(m.olr) # assuming normality
```

By exponentiating the coefficients, we obtain odds ratios.

For every one unit increase in median clr-transformed abundance, the odds of the reappearance count being greater is multiplied by 2.192, holding the log sample read depth constant.

```{r}
exp(cbind(oddsRatio = coef(m.olr), olr.profiled.ci))
```

Yeah, not a great prediction. Very 1-heavy.
```{r}
summary(predict(m.olr, positive.reappearances.f4))
```

What about an interaction betwee median abundance and log sample read depth? 
median abundance is not so significant a predictor.
Again, the predictions are not great.
```{r}
olr.interact <- polr(y.factor ~ median.abundance*log(sample.read.depth), 
                     data = positive.reappearances.f4, Hess = TRUE)

summary(olr.interact)

olr.interact.ctable <- coef(summary(olr.interact))

olr.interact.pvals <- pnorm(abs(olr.interact.ctable[ , "t value"]), lower.tail = FALSE) * 2
(olr.interact.ctable <- cbind(olr.interact.ctable, "p value" = olr.interact.pvals))
(olr.interact.profiled.ci <- confint(olr.interact)) # profiled CIs
confint.default(olr.interact) # assuming normality

summary(predict(olr.interact, positive.reappearances.f4))
```





